pkg/storage/chunk/chunk.go

```diff
@@ -89,18 +88,26 @@ func NewChunk(userID string, fp model.Fingerprint, metric labels.Labels, c prom_
 // Post-checksums, externals keys become the same across DynamoDB, Memcache
 // and S3.  Numbers become hex encoded.  Keys look like:
 // `<user id>/<fingerprint>:<start time>:<end time>:<checksum>`.
+//
+// Post-v12, (2) additional prefixes were added to external keys
+// to support better read and write request parallelization:
+// `<user>/<period>/<shard>/<fprint>/<start>:<end>:<checksum>`
 func ParseExternalKey(userID, externalKey string) (Chunk, error) {
```

Comment:
@jeschkies raised a great question about this function in an earlier iteration of this PR:

> I've been there. May I suggest to add a key version prefix or suffix? E.g.
> 
> `keyv1:<user id>/<fingerprint>:<start time>:<end time>:<checksum>`
> 
> Unless you are really sure that the key schema will never change again.

I think this is a fair point as right now our process for parsing relies on the structure of the key as opposed to a specific identifier to understand how to proceed.

Now that we've separated out the key construction from the parsing (we did this so that there was less inter-dependence between the ObjectClient and the Schema), we considered what a refactor would look like if we wanted to introduce _yet another key structure_. It seems that either way we would either be adding, removing, or reordering prefixes which all would require changes to this conditional and our associated parsing functions.

Because of this, routing like we are in this function seems to make the most sense but I'm open to comment here if there is a better way.